compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDistributeCoresAndObjectFifos.cpp

```diff
@@ -388,88 +638,252 @@ class AssignAieTiles
+    if (!targetTiles.empty() && !sourceTiles.empty()) {
+      return logicalObjectFifo.emitOpError()
+             << "found logical objectfifo with both source and target tiles, "
+                "which is not supported yet";
+    } else if (!targetTiles.empty()) {
+      // Create tile locations for this logical objectfifo based on target
+      // tiles.
+      if (failed(createTileLocations(targetTiles))) {
+        return failure();
+      }
+    } else if (!sourceTiles.empty()) {
+      // Create tile locations for this logical objectfifo based on source
+      // tiles.
+      if (failed(createTileLocations(sourceTiles))) {
         return failure();
       }
     } else {
-      return logicalObjectFifo.emitOpError()
-             << "found logical objectfifo with unknown memory space";
+      // Don't assign this logicalObjectFifo to a physical tile (yet!). Wait
+      // for other logical objectfifos to be assigned first.
+      return failure();
     }
+
     // If no tile results, skip, and maybe in a next iteration another tile will
     // be found.
-    if (tileResults.empty()) {
+    if (tileLocations.empty()) {
       return failure();
     }
 
-    // Extend this logical objectfifo's tile set.
-    SmallVector<Value> objFifoTiles = logicalObjectFifo.getTiles();
-    DenseSet<Value> tileSet(objFifoTiles.begin(), objFifoTiles.end());
+    rewriter.setInsertionPoint(logicalObjectFifo);
+    rewriter.replaceOpWithNewOp<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+        logicalObjectFifo, logicalObjectFifo.getMemref(),
+        tileLocations.takeVector());
+    return success();
+  }
+};
 
-    // If the logical objectfifo already contains all the new tiles, skip.
-    if (llvm::all_of(tileResults,
-                     [&](Value val) { return tileSet.contains(val); })) {
-      return failure();
+/// Return the user DMA operations and corresponding assigned tiles in the
+/// specified direction (source or target).
+template <CopyOpOperateOn OperateOn>
+SmallVector<std::pair<AMDAIE::DmaCpyNdOp, SmallVector<AMDAIE::TileOp>>>
+getUserDmasAndTiles(AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjectFifo) {
+  SmallVector<std::pair<AMDAIE::DmaCpyNdOp, SmallVector<AMDAIE::TileOp>>>
+      dmaOps;
+  for (Operation *user : logicalObjectFifo->getUsers()) {
+    if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(user)) {
+      SmallVector<AMDAIE::TileOp> tiles;
+      ValueRange tileIndices;
+      if constexpr (OperateOn == CopyOpOperateOn::Source) {
+        if (dmaOp.getTargetObjectFifo() != logicalObjectFifo) continue;
+        tileIndices = dmaOp.getSourceObjectFifo().getTiles();
+      } else if constexpr (OperateOn == CopyOpOperateOn::Target) {
+        if (dmaOp.getSourceObjectFifo() != logicalObjectFifo) continue;
+        tileIndices = dmaOp.getTargetObjectFifo().getTiles();
+      }
+      for (Value index : tileIndices)
+        tiles.push_back(dyn_cast<AMDAIE::TileOp>(index.getDefiningOp()));
```

Comment:
Is it needed to go through all the indices in tileIndices? This will result in duplicate tiles? (And maybe it's the answer of my previous comment...)

Reply:
Yes, there can be multiple target tiles through different DMAs resulting in duplicates. Here, we're just collecting all possible ones.